Product controller refactoring

yuya-yuliya opened the pull request:
- replace ProductController with ProductSearchApiController
- remove Get product endpoint
- ICatalogRepository and I ProductListRepository are obsolete
- add ICatalogService and IProductListService
- product models refactoring

larionovadaria commented on src/Feature/Catalog/code/Controllers/ProductsSearchApiController.cs:
```diff
@@ -0,0 +1,122 @@
+        public ProductsSearchApiController(
+            IProductListService productListService)
+        {
+            this.productListService = productListService;
```
Assert

larionovadaria commented on src/Feature/Catalog/code/Controllers/ProductsSearchApiController.cs:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        [ResponseType(typeof(ProductListResultModel))]
+        [HttpGet]
```
will  `[HttpGet, Route]` work with empty routes?

yuya-yuliya replied:
Yes

larionovadaria commented on src/Feature/Catalog/code/Controllers/ProductsSearchApiController.cs:
```diff
@@ -0,0 +1,122 @@
+            };
+        }
+
+        private IHttpActionResult Execute<T>(Func<Result<T>> action)
```
- add todo
- add task for investigating how to avoid this code duplication

for now left as is

larionovadaria commented on src/Foundation/Commerce/code/Models/Catalog/BaseProductModel.cs:
```diff
@@ -14,36 +14,37 @@
 
     public class BaseProductModel
     {
+        [Obsolete("Use BaseProductModel(BaseProduct, ICurrencyProvider)")]
```
BaseProductModel was renamed to BaseProduct. Take a look on that.

larionovadaria commented on src/Foundation/Commerce/code/Models/Catalog/ProductModel.cs:
```diff
@@ -14,15 +14,21 @@
     [TsClass]
     public class ProductModel : BaseProductModel
     {
+        [Obsolete("Use ProductModel(Product, ICurrencyProvider)")]
```
Same. Renamed to Product

larionovadaria commented on src/Foundation/Commerce/code/Models/Catalog/ProductVariantModel.cs:
```diff
@@ -14,36 +14,55 @@
     [TsClass]
     public class ProductVariantModel : BaseProductModel
     {
+        [Obsolete("Use ProductVariantModel(Variant, ICurrencyProvider)")]
```
Renamed to ProductVariant

larionovadaria commented on src/Foundation/Commerce/code/Models/Catalog/ProductsSearchParametersModel.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    using System.Collections.Specialized;
+
+    public class ProductsSearchParametersModel
```
Don't use Model postfix. 
Rename to `ProductSearchOptions`

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/CatalogService.cs:
```diff
@@ -0,0 +1,155 @@
+            this.searchManager = searchManager;
+        }
+
+        public Result<ProductModel> GetProduct(string productId)
```
ProductModel => Product

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/CatalogService.cs:
```diff
@@ -0,0 +1,155 @@
+                    });
+        }
+
+        public Result<ProductModel> GetProduct(Item productItem)
```
public?

Commit:
Product controller refactoring

- add SearchController
- add ICatalogService and IProductListService
- product models refactoring
- add IMapper and Mapper
- add ICatalogEntityMapper and CatalogEntityMapper
- add unit tests

larionovadaria commented on src/Feature/Catalog/code/ModelMappers/ICatalogEntityMapper.cs:
```diff
@@ -0,0 +1,33 @@
+    using Models.Requests;
+
+    /// <summary>
+    /// Perform mapping for catalog models
```
performs

larionovadaria commented on src/Feature/Catalog/code/ModelMappers/ICatalogEntityMapper.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        /// <param name="searchRequest">Search request</param>
+        /// <returns>Products search options</returns>
+        ProductsSearchOptions MapToProductSearchOptions(ProductSearchRequest searchRequest);
```
ProductSearchOptions

larionovadaria commented on src/Feature/Catalog/code/Models/MappingProfiles/ProductSearchRequestMappingProfile.cs:
```diff
@@ -0,0 +1,49 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+namespace Wooli.Feature.Catalog.Models.MappingProfiles
```
Models/Mappings/

larionovadaria commented on src/Feature/Catalog/code/Models/MappingProfiles/ProductSearchRequestMappingProfile.cs:
```diff
@@ -0,0 +1,49 @@
+        public ProductSearchRequestMappingProfile()
+        {
+            this.CreateMap<ProductSearchRequest, ProductsSearchOptions>()
+                .ForMember(model => model.SearchKeyword, opt => opt.MapFrom(dto => dto.SearchKeyword))
```
No model and dtos, only source and destinations

larionovadaria commented on src/Feature/Catalog/code/Models/MappingProfiles/ProductSearchRequestMappingProfile.cs:
```diff
@@ -0,0 +1,49 @@
+                .ForMember(model => model.SortDirection, opt => opt.MapFrom(dto => dto.SortDirection))
+                .ForMember(model => model.CurrentCatalogItemId, opt => opt.MapFrom(dto => dto.CurrentCatalogItemId))
+                .ForMember(model => model.CurrentItemId, opt => opt.MapFrom(dto => dto.CurrentItemId))
+                .ForAllOtherMembers(opt => opt.Ignore());
```
Why?

larionovadaria commented on src/Feature/Catalog/code/Wooli.Feature.Catalog.csproj:
```diff
@@ -62,6 +62,9 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="AutoMapper, Version=6.2.2.0, Culture=neutral, PublicKeyToken=be96cd2c38ef1005, processorArchitecture=MSIL">
```
Please make sure this version is used in all other projects.

yuya-yuliya replied:
Yes, we use 6.2.2 version in our other projects

larionovadaria commented on src/Feature/Catalog/tests/Wooli.Feature.Catalog.Tests.csproj:
```diff
@@ -526,6 +527,20 @@
+      <Project>{b4cd14e8-725d-401c-8d99-f20ef83e3ff5}</Project>
+      <Name>Wooli.Foundation.Extensions</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\code\Wooli.Feature.Catalog.csproj">
```
Feature to Feature? No way.

larionovadaria commented on src/Foundation/Commerce/code/Models/Entities/FacetResult.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public string Name { get; set; }
+
+        public void Initialize(CommerceQueryFacet commerceQueryFacet)
```
No initialization here. All the logic should be moved to mappers/services etc.

larionovadaria commented on src/Foundation/Commerce/code/Models/Entities/FacetValueResult.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public string Name { get; protected set; }
+
+        public void Initialize(FacetValue queryFacet)
```
Same

larionovadaria commented on src/Foundation/Commerce/code/Models/Entities/FacetValueResult.cs:
```diff
@@ -0,0 +1,37 @@
+
+    using TypeLite;
+
+    [TsClass]
```
Exclude from code coverage. And check all the new models for that attribute

larionovadaria commented on src/Foundation/Commerce/code/Models/Entities/ProductListResult.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public int TotalItemCount { get; set; }
+
+        public void Initialize(
```
Same

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/ICatalogService.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <summary>
+    /// Performs operations with catalog
+    /// </summary>
+    public interface ICatalogService
```
Left only this interface, let's implementation will be with "not implemented exception" for now

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/ICatalogService.cs:
```diff
@@ -0,0 +1,45 @@
+        /// Gets current product
+        /// </summary>
+        /// <returns>Product result</returns>
+        Result<Product> GetCurrentProduct();
```
Current for what?

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/ICatalogService.cs:
```diff
@@ -0,0 +1,45 @@
+        /// Gets current category
+        /// </summary>
+        /// <returns>Product result</returns>
+        Result<Category> GetCurrentCategory();
```
Current for what?

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/IProductListService.cs:
```diff
@@ -0,0 +1,32 @@
+        /// </summary>
+        /// <param name="options">Search parameters object</param>
+        /// <returns>List of products</returns>
+        Result<ProductListResult> GetProductList(ProductsSearchOptions options);
```
Left implementation with "not implemented exception"

stanislav-kozorez commented on src/Feature/Catalog/code/Controllers/SearchController.cs:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        // TODO: Create Extension Method
+        private ActionResult Execute<T>(Func<Result<T>> action)
```
`BaseController.Execute` can already be used

larionovadaria commented on src/Foundation/Base/code/ModelMappers/IMapper.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Wooli.Foundation.Base.ModelMappers
```
- Please move to "Mappers" folder 
- Also, please update Guidelines with description of these changes and how to use it

yuya-yuliya replied:
I'll update guidelines after merge

larionovadaria commented on src/Foundation/Base/code/ModelMappers/Mapper.cs:
```diff
@@ -0,0 +1,41 @@
+
+    public class Mapper : IMapper
+    {
+        protected readonly AutoMapper.IMapper InnerMapper;
```
innerMapper

larionovadaria commented on src/Foundation/Base/code/ModelMappers/Mapper.cs:
```diff
@@ -0,0 +1,41 @@
+        private MapperConfigurationExpression GetMappings(Assembly assembly)
+        {
+            var baseMappings = new MapperConfigurationExpression();
+            baseMappings.AddProfiles(assembly);
```
We discussed with you issue with auto-registration and DI here. We can also separate it as we want, like it already done in Foundation.Commerce
```
public abstract class Mapper
    {
        protected abstract readonly IMapper mapper;
        protected abstract MapperConfiguration configuration {get;set;}

        public Mapper()
        { 
        }

        public TDestination Map<TSource, TDestination>(TSource source)
        {
            return this.mapper.Map<TSource, TDestination>(source);
        }
    }
```
```
public CategoryMapper : Mapper, ICategoryMapper
{
            ///...
         MapperConfiguration configuration = new ...
         this.mapper = new Mapper(configuration);
}
```
something like that, no profiles

larionovadaria commented on src/Feature/Catalog/code/Controllers/SearchController.cs:
```diff
@@ -0,0 +1,53 @@
+    using Foundation.Commerce.Models.Entities;
+    using Foundation.Commerce.Services.Catalog;
+
+    using ModelMappers;
```
Mappers

larionovadaria commented on src/Feature/Catalog/code/ModelMappers/CatalogEntityMapper.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            var configuration = new MapperConfiguration(cfg => cfg.AddProfiles(Assembly.GetExecutingAssembly()));
+
+            this.InnerMapper = new AutoMapper.Mapper(configuration);
```
this.innerMapper

yuya-yuliya replied:
We have Upper case naming of protected fields across the solution.

yuya-yuliya replied:
We have Upper case naming of protected fields across the solution.

larionovadaria commented on src/Feature/Catalog/code/ModelMappers/CatalogEntityMapper.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        public CatalogEntityMapper()
+        {
+            var configuration = new MapperConfiguration(cfg => cfg.AddProfiles(Assembly.GetExecutingAssembly()));
```
You can specify mapping right here, because definitely there will be no more mappings in this class

larionovadaria commented on src/Feature/Catalog/tests/Controllers/SearchControllerTests.cs:
```diff
@@ -0,0 +1,60 @@
+    using Foundation.Commerce.Models;
+    using Foundation.Commerce.Models.Entities;
+
+    using ModelMappers;
```
Mappers

larionovadaria commented on src/Feature/Catalog/tests/Controllers/SearchControllerTests.cs:
```diff
@@ -0,0 +1,60 @@
+
+    public class SearchControllerTests
+    {
+        private readonly IProductSearchService productSearchService;
```
Why there's new line between private fields?

yuya-yuliya replied:
Because this fields are logically different, so we split them by empty line to show this difference in formatting.

larionovadaria commented on src/Foundation/Base/code/Mappers/Mapper.cs:
```diff
@@ -0,0 +1,16 @@
+
+    public abstract class Mapper : IMapper
+    {
+        protected AutoMapper.IMapper InnerMapper { get; set; }
```
innerMapper

yuya-yuliya replied:
We have Upper case naming of protected fields across the solution.

larionovadaria commented on src/Foundation/Base/code/Mappers/Mapper.cs:
```diff
@@ -0,0 +1,16 @@
+
+        public TResult Map<TSource, TResult>(TSource source)
+        {
+            Assert.IsNotNull(this.InnerMapper, "InnerMapper must be initialized");
```
also needed assert for input parameter

larionovadaria commented on src/Foundation/Commerce/code/Models/Entities/FacetResult.cs:
```diff
@@ -0,0 +1,35 @@
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+
+    using Sitecore.Commerce.Engine.Connect.Search.Models;
```
Why do we need here this using? We have our own model FacetValueResult

larionovadaria commented on src/Foundation/Commerce/code/Services/Catalog/IProductSearchService.cs:
```diff
@@ -0,0 +1,32 @@
+    using Models.Entities;
+
+    /// <summary>
+    /// Perform main operations with catalog
```
Performs
